pandas/core/window/aggregators.py

```diff
@@ -0,0 +1,159 @@
+
+    def __init__(self):
+        self.count = 0
+
+    @abc.abstractmethod
+    def finalize(self):
+        """Return the final value of the aggregation."""
+
+    @classmethod
+    @abc.abstractmethod
+    def make_aggregator(
+        cls, values: np.ndarray, minimum_periods: int
+    ) -> BaseAggregator:
+        """Return an aggregator that performs the aggregation calculation"""
+
+
+class UnaryAggKernel(AggKernel):
+    @abc.abstractmethod
+    def step(self, value) -> None:
+        """Update the state of the aggregation with `value`."""
+
+
+class SubtractableAggregator(BaseAggregator):
+    def __init__(self, values: np.ndarray, min_periods: int, agg: AggKernel) -> None:
+        super().__init__(values, min_periods)
+        self.agg = agg
+        self.previous_start = -1
+        self.previous_end = -1
+
+    def query(self, start: int, stop: int) -> Optional[Scalar]:
+        """Compute a value based on changes in bounds."""
+        if self.previous_start == -1 and self.previous_end == -1:
+            for value in self.values[start:stop]:
+                self.agg.step(value)
+        else:
+            for value in self.values[self.previous_start : start]:
+                self.agg.invert(value)
+            for value in self.values[self.previous_end : stop]:
+                self.agg.step(value)
+        self.previous_start = start
+        self.previous_end = stop
+        if self._meets_minimum_periods(self.values[start:stop]):
+            return self.agg.finalize()
+        return None
+
+
+class Sum(UnaryAggKernel):
+    def __init__(self) -> None:
+        super().__init__()
+        self.total = 0
+
+    def step(self, value) -> None:
+        if not np.isnan(value):
+            self.count += 1
+            self.total += value
+
+    def invert(self, value) -> None:
+        """Used only in subtractable kernels."""
+        if not np.isnan(value):
+            self.count -= 1
+            self.total -= value
+
+    def finalize(self) -> Optional[int]:
+        if not self.count:
+            return None
+        return self.total
+
+    def combine(self, other) -> None:
+        """Used only in segment tree aggregator."""
+        self.total += other.total
+        self.count += other.count
+
+    @classmethod
+    def make_aggregator(cls, values: np.ndarray, min_periods: int) -> BaseAggregator:
+        aggregator = SubtractableAggregator(values, min_periods, cls())
+        return aggregator
+
+
+class Mean(Sum):
+    def finalize(self) -> Optional[float]:
+        if not self.count:
+            return None
+        return self.total / self.count
+
+
+def rolling_aggregation(
+    values: np.ndarray,
+    begin: np.ndarray,
+    end: np.ndarray,
+    minimum_periods: int,
+    kernel_class,
+) -> np.ndarray:
+    """Perform a generic rolling aggregation"""
+    aggregator = kernel_class.make_aggregator(values, minimum_periods)
+    result = np.empty(len(begin))
```

Comment:
how does this know the result dtype?

Reply:
We might need to make this an attribute of a Kernel because the aggregation will dictate the dtype (e.g. `int` for `Sum`, `float` for `Mean`).

Since `np.empty` defaults to `float` and mean should return floats, we are okay here for now.